runtime/cudaq/dynamics/cudm_helpers.cpp

```diff
@@ -0,0 +1,209 @@
+  }
+  return dimensions;
+}
+
+cudensitymatOperator_t convert_to_cudensitymat_operator(
+    cudensitymatHandle_t handle,
+    const std::map<std::string, std::complex<double>> &parameters,
+    const operator_sum &op, const std::vector<int64_t> &mode_extents) {
+  try {
+    cudensitymatOperator_t operator_handle;
+    auto status = cudensitymatCreateOperator(
+        handle, static_cast<int32_t>(mode_extents.size()), mode_extents.data(),
+        &operator_handle);
+    if (status != CUDENSITYMAT_STATUS_SUCCESS) {
+      throw std::runtime_error("Failed to create operator.");
+    }
+
+    for (const auto &product_op : op.get_terms()) {
+      cudensitymatOperatorTerm_t term;
+
+      HANDLE_CUDM_ERROR(cudensitymatCreateOperatorTerm(
+          handle, static_cast<int32_t>(mode_extents.size()),
+          mode_extents.data(), &term));
+
+      for (const auto &component : product_op.get_terms()) {
+        if (std::holds_alternative<cudaq::elementary_operator>(component)) {
+          const auto &elem_op = std::get<cudaq::elementary_operator>(component);
+
+          // Create a cudensitymat elementary operator
+          cudensitymatElementaryOperator_t cudm_elem_op;
+
+          // Get the matrix representation of elementary operator
+          auto dimensions = convert_dimensions(mode_extents);
+          auto matrix = elem_op.to_matrix(dimensions, parameters);
+
+          // Flatten the matrix into a single-dimensional array
+          std::vector<std::complex<double>> flat_matrix;
+          for (size_t i = 0; i < matrix.get_rows(); i++) {
+            for (size_t j = 0; j < matrix.get_columns(); j++) {
+              flat_matrix.push_back(matrix[{i, j}]);
+            }
+          }
+
+          // Create a cudensitymat elementary operator
+          HANDLE_CUDM_ERROR(cudensitymatCreateElementaryOperator(
+              handle, 1, mode_extents.data(),
+              CUDENSITYMAT_OPERATOR_SPARSITY_NONE, 0, nullptr, CUDA_C_64F,
+              flat_matrix.data(), {nullptr, nullptr}, &cudm_elem_op));
+
+          // Append the elementary operator to the term
+          HANDLE_CUDM_ERROR(cudensitymatOperatorTermAppendElementaryProduct(
+              handle, term, 1, &cudm_elem_op, &elem_op.degrees[0], nullptr,
+              make_cuDoubleComplex(1.0, 0.0), {nullptr, nullptr}));
+
+          // Destroy the elementary operator after appending
+          HANDLE_CUDM_ERROR(
+              cudensitymatDestroyElementaryOperator(cudm_elem_op));
+        } else if (std::holds_alternative<cudaq::scalar_operator>(component)) {
+          const auto &scalar_op = std::get<cudaq::scalar_operator>(component);
+
+          // Use the scalar coefficient
+          auto coeff = scalar_op.evaluate(parameters);
+          HANDLE_CUDM_ERROR(cudensitymatOperatorTermAppendElementaryProduct(
+              handle, term, 0, nullptr, nullptr, nullptr,
+              {make_cuDoubleComplex(coeff.real(), coeff.imag())},
+              {nullptr, nullptr}));
```

Comment:
We'll need to add a `TODO` here to handle callback scalars (time-dependent) via `cudensitymatScalarCallback_t`.